apps/raulpe7eira/exercises/pure_otp/lib/todo.ex

```diff
@@ -0,0 +1,152 @@
+  end
+
+  @doc """
+  Completes a created Task.
+  """
+  def complete(id_task) do
+    send_request :complete, self(), [id_task]
+  end
+
+  @doc """
+  Shuts this control down, for force the supervisor the restart it.
+  """
+  def shutdown do
+    send_request :shutdown, self()
+  end
+
+  defp send_request(action, caller) do
+    send __MODULE__, {action, caller}
+    receive_response action
+  end
+  defp send_request(action, caller, params) do
+    send __MODULE__, {action, caller, params}
+    receive_response action
+  end
+
+  defp receive_response(action) do
+    receive do
+      {:ok, value} -> render_response({action, value})
+      {:error, value} -> treat_error value
+    after @timeout_ms -> treat_error "ops... I forgot to #{action}"
+    end
+  end
+
+  defp render_response({:add, task}) when is_map(task) do
+    task
+    |> filter_by_keys([:completed_at])
+  end
+  defp render_response({:list, todo}) when is_list(todo) do
+    todo
+    |> Enum.map(&(filter_by_keys(&1, [:created_at, :completed_at])))
+  end
+  defp render_response({:complete, task}) when is_map(task) do
+    task
+    |> filter_by_keys([])
+  end
+  defp render_response({:shutdown, message}) do
+    treat_success message
+  end
+
+  defp filter_by_keys(task, filter_keys) when is_map(task) and is_list(filter_keys) do
+    task
+    |> Map.from_struct
+    |> Enum.filter(fn {key, _value} -> key not in filter_keys end)
+    |> Enum.into(%{})
+  end
+
+  defp treat_success(message) do
+    %{ok: message}
+  end
+
+  defp treat_error(reason) do
+    %{error: reason}
+  end
+
+  @doc """
+  Receiving a request, while this control is started.
+  """
+  def receiving_requests(todo) when is_list(todo) do
+    receive do
+      {:add, caller, [title, completed]} ->
+        task = Task.create todo, title, completed
+        send_response caller, task
+        todo = insert todo, task
+        receiving_requests todo
+
+      {:list, caller} ->
+        send_response caller, todo
+        receiving_requests todo
+
+      {:complete, caller, [id_task]} ->
+        task = Task.complete todo, id_task
+        todo = update todo, task
+        send_response caller, task
+        receiving_requests todo
+
+      {:shutdown, caller} ->
+        send_response caller, "TODO shutdown"
+    end
+  end
+
+  defp insert(todo, task) when is_list(todo) and is_map(task) do
+    [task | todo]
+  end
+  defp insert(todo, _task) when is_list(todo), do: todo 
+
+  defp update(todo, task) when is_list(todo) and is_map(task) do
+    todo
+    |> Enum.map(&(if &1.id == task.id, do: task, else: &1))
+  end
+  defp update(todo, _task) when is_list(todo), do: todo
+
+  defp send_response(caller, {:unchecked, reason}), do: send caller, {:error, reason}
+  defp send_response(caller, task) when is_map(task), do: send caller, {:ok, task}
+  defp send_response(caller, todo) when is_list(todo), do: send caller, {:ok, todo}
+  defp send_response(caller, message), do: send caller, {:ok, message}
```

Comment:
Fez bastante uso de pattern matcher, ficou bacana... `código sem if` :heart:

Reply:
To gostando bastante dessa abordagem no código, visualmente vc cria blocos de responsabilidade menores, mas é um exercício, ainda mais para quem veio de java onde tudo era uma bíblia.